### PR TITLE
Add test case: failed backups cleanup

### PR DESCRIPTION
**Pull request by mantissahz**
longhorn/longhorn#4481

Signed-off-by: James Lu <james.lu@suse.com>

### CODE REVIEW

File: manager/integration/tests/test_basic.py [hunk 1]
```diff
@@ -4438,3 +4439,87 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+        return b.id == backup_id and "Error" in b.state and b.error != ""
+    vol = common.wait_for_backup_state(client, volume_name,
+                                       backup_failure_predicate)
+
+    # wait in 5 minutes for automatic failed backup cleanup
+    wait_for_backup_delete(client, volume_name, backup.name)
+
+    # cleanup and reset
+    delete_backup_volume(client, volume_name)
+    cleanup_volume(client, vol)
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:4524]: nit: already handled in [cleanup_client](https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/tests/common.py#L1478-L1505).

File: manager/integration/tests/test_basic.py [hunk 2]
```diff
@@ -4438,3 +4439,87 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+        failed_backup_ttl_setting, value=failed_backup_ttl)
+    assert new_failed_backup_ttl_setting.value == failed_backup_ttl
+
+    backupstore_cleanup(client)
+
+    # create a volume
+    vol = create_and_check_volume(client, volume_name,
+                                  num_of_replicas=2, size=size)
+
+    # attatch the volume
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:4483]: attach

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: manager/integration/tests/test_basic.py [hunk 3]
```diff
@@ -4438,3 +4439,87 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
 
     """
     pass
+
+
+@pytest.mark.coretest  # NOQA
+def test_backup_failed_auto_cleanup(set_random_backupstore,  # NOQA
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py:4445]: OK, I will add the another one to cover disabling the auto cleanup

File: manager/integration/tests/test_basic.py (hunk 2 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py (hunk 1 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Removed. Thanks.

File: manager/integration/tests/test_basic.py [hunk 4]
```diff
@@ -4438,3 +4439,87 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+                                    client, core_api, volume_name):  # NOQA
+    """
+    Test the failed backup would be automatically deleted.
+
+    1.   Set the default setting `backupstore-poll-interval` to 60 (seconds)
+    2.   Set the defualt setting `failed-backup-ttl` to 3 (minutes)
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py [hunk 5]
```diff
@@ -4438,3 +4439,87 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+    3.   Create a volume and attach to the current node
+    4.   Create a empty backup for creating the backup volume
+    5.   Write some data to the volume
+    6.   Create a backup of the volume
+    7.   Crash all replicas
+    8.   Wait and check the backup failed
+    9.   Wait and check if the backup was deleted automatically.
+    10.  Cleanup
+    """
+    failed_backup_auto_cleanup(client, core_api, volume_name, size=str(256 * Mi))
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Removed, just for testing. Thanks.

File: manager/integration/tests/test_basic.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py:4445]: Add case 2: disable auto-cleanup after auto-cleanup succeeded.

File: manager/integration/tests/test_basic.py [hunk 6]
```diff
@@ -4438,3 +4439,128 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+    # create a snapshot and a backup
+    snap = create_snapshot(client, volume_name)
+    vol = client.by_id_volume(volume_name)
+    vol.snapshotBackup(name=snap.name)
+
+    # check backup status is in an InProgress state
+    _, backup = find_backup(client, volume_name, snap.name)
+    backup_id = backup.id
+
+    def backup_inprogress_predicate(b):
```
Comment (innobead) [inline manager/integration/tests/test_basic.py:4546]: Can we separate this to two test cases/methods instead? you can use one as a basic test method, but the other just base on it.

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, but one comment required to fix. 

Also, the sign-off is not right, should use your SUSE email.

File: manager/integration/tests/test_basic.py [hunk 7]
```diff
@@ -4438,3 +4439,128 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+    Test the failed backup would be automatically deleted.
+
+    Prepare:
+    1.   Set the default setting `backupstore-poll-interval` to 60 (seconds)
+    2.   Set the default setting `failed-backup-ttl` to 3 (minutes)
+    3.   Create a volume and attach to the current node
+    4.   Create a empty backup for creating the backup volume
+    5.   Write some data to the volume
+
+    Case 1: enable automatic failed backupcleanup
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:4457]:
```suggestion
    Case 1: enable automatic failed backup cleanup
```

File: manager/integration/tests/test_basic.py [hunk 8]
```diff
@@ -4438,3 +4439,128 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+    1.   Create a backup of the volume
+    2.   Crash all replicas
+    3.   Wait and check the backup failed
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:4460]:
```suggestion
    3.   Wait and check if the backup failed
```

File: manager/integration/tests/test_basic.py [hunk 9]
```diff
@@ -4438,3 +4439,128 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+    4.   Wait and check if the backup was deleted automatically
+
+    Case 2: Disable automatic failed backup cleanup
+    1.   Disable the default setting `failed-backup-ttl` to 0
+    2.   Create a backup of the volume
+    3.   Crash all replicas
+    4.   Wait and check the backup failed
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:4467]:
```suggestion
    4.   Wait and check if the backup failed
```

File: manager/integration/tests/test_basic.py [hunk 10]
```diff
@@ -4438,3 +4439,128 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+    # backup status should be in an Error state and with an error message
+    def backup_failure_predicate(b):
+        return b.id == backup_id and "Error" in b.state and b.error != ""
+    common.wait_for_backup_state(client, volume_name,
+                                 backup_failure_predicate)
+
+    # wait for 5 minutes to check failed backup exists
+    try:
+        wait_for_backup_delete(client, volume_name, backup.name)
+        assert False, "deleted backup " + backup.name + " for volume " \
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:4563]:
```suggestion
        assert False, "backup " + backup.name + " for volume " \
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: manager/integration/tests/test_basic.py (hunk 6 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py:4537]: Do you mean the enable and disable cases?

Corrected the sign-off email, thanks.

File: manager/integration/tests/test_basic.py (hunk 7 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Removed. Thanks.

File: manager/integration/tests/test_basic.py (hunk 8 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py (hunk 9 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py (hunk 10 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py (hunk 6 above)
Comment (c3y1huang) [reply manager/integration/tests/test_basic.py:4537]: The code look duplicated. You can have a function(for example, backup_disable_auto_cleanup_test) to cover the duplicated code and use it in both cases.

Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Yeah, modified, moved the duplicated code into a method `backup_failed_cleanup` and both enabling and disabling test case use the method.
Thanks.

File: manager/integration/tests/test_basic.py [hunk 11]
```diff
@@ -4438,3 +4439,122 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
...
+    # backup status should be in an Error state and with an error message
+    def backup_failure_predicate(b):
+        return b.id == backup_id and "Error" in b.state and b.error != ""
+    common.wait_for_backup_state(client, volume_name,
+                                 backup_failure_predicate)
+
+    return backup_name
+
+
+@pytest.mark.coretest  # NOQA
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:4511]: Do we want the cases in the `coretest`?

cc @longhorn/qa

**Review by c3y1huang — APPROVED** [review]
LGTM

File: manager/integration/tests/test_basic.py [hunk 12]
```diff
@@ -4438,3 +4439,122 @@ def test_space_usage_for_rebuilding_only_volume():  # NOQA
 
     """
     pass
+
+
+def backup_failed_cleanup(client, core_api, volume_name, volume_size,  # NOQA
```
Comment (innobead) [inline manager/integration/tests/test_basic.py:4444]: nit: would suggest having volume_size fixed to ensure the following status check is always correct.

**Review by innobead — APPROVED** [review]
Please backport this to 1.2.x.

**Pull request merged**